File: src/main/services/ModArchivePlanner.ts
import fs from 'fs';
import path from 'path';
import type { ModInstallPlanDto } from '../../shared/types';
import { GMP_METADATA_FOLDER, GMP_FALLBACK_METADATA_FOLDER, TEMP_FOLDER } from '../constants';
const PLUGIN_EXT = '.dll';
export interface InstallTargets {
    modded: {
        src: string;
        destRel: string;
    }[];
    plugins: {
        src: string;
        destRel: string;
        extras: string[];
    }[];
    patchers: {
        src: string;
        destRel: string;
    }[];
    modInfo: {
        src: string;
        destRel: string;
    }[];
}
function pluginsRootCandidates(extractRoot: string): string[] {
    return [path.join(extractRoot, 'BepInEx', 'plugins'), path.join(extractRoot, 'plugins')];
}
function moddedRootCandidates(extractRoot: string): string[] {
    return [
        path.join(extractRoot, 'BALDI_Data', 'StreamingAssets', 'Modded'),
        path.join(extractRoot, 'Modded')
    ];
}
function patchersRootCandidates(extractRoot: string): string[] {
    return [path.join(extractRoot, 'BepInEx', 'patchers'), path.join(extractRoot, 'patchers')];
}
function pluginTarget(absDll: string, rel: string): {
    src: string;
    destRel: string;
    extras: string[];
} {
    const dir = path.dirname(absDll);
    const stem = path.basename(absDll, PLUGIN_EXT);
    const extras = [] as string[];
    for (const f of ['xml', 'pdb']) {
        const p = path.join(dir, `${stem}.${f}`);
        if (fs.existsSync(p))
            extras.push(p);
    }
    return { src: absDll, destRel: rel, extras };
}
function walkDlls(dir: string, rel: string, into: {
    src: string;
    destRel: string;
    extras: string[];
}[]): void {
    for (const e of fs.readdirSync(dir, { withFileTypes: true })) {
        if (e.isDirectory()) {
            walkDlls(path.join(dir, e.name), rel ? `${rel}/${e.name}` : e.name, into);
        }
        else if (e.name.toLowerCase().endsWith(PLUGIN_EXT)) {
            const r = rel ? `${rel}/${e.name}` : e.name;
            into.push(pluginTarget(path.join(dir, e.name), r));
        }
    }
}
function walkPatchers(dir: string, rel: string, into: {
    src: string;
    destRel: string;
}[]): void {
    for (const e of fs.readdirSync(dir, { withFileTypes: true })) {
        if (e.isDirectory()) {
            walkPatchers(path.join(dir, e.name), rel ? `${rel}/${e.name}` : e.name, into);
        }
        else if (e.name.toLowerCase().endsWith(PLUGIN_EXT)) {
            const r = rel ? `${rel}/${e.name}` : e.name;
            into.push({ src: path.join(dir, e.name), destRel: r });
        }
    }
}
export function isValidGuidFolderName(input: string): boolean {
    if (!input)
        return false;
    if (!/^[a-z0-9.]+$/.test(input))
        return false;
    const segs = input.split('.');
    return segs.length >= 2 && segs.length <= 5;
}
export function isTemplateFolderName(input: string): boolean {
    const n = input.toLowerCase();
    return n.includes('template') || n.includes('example');
}
function relDirOf(rel: string): string {
    const i = rel.lastIndexOf('/');
    return i === -1 ? '' : rel.slice(0, i);
}
function stemOf(rel: string): string {
    const base = path.basename(rel);
    const dot = base.lastIndexOf('.');
    return dot === -1 ? base : base.slice(0, dot);
}
function findPluginEntry(plugins: InstallTargets['plugins'], rel: string, stem: string): {
    src: string;
    destRel: string;
    extras: string[];
} | undefined {
    const dir = relDirOf(rel);
    return (plugins.find((p) => relDirOf(p.destRel) === dir && stemOf(p.destRel).toLowerCase() === stem) ??
        plugins.find((p) => stemOf(p.destRel).toLowerCase() === stem));
}
function walkHeuristic(extractRoot: string, targets: InstallTargets): void {
    const skipNames = new Set([GMP_METADATA_FOLDER, GMP_FALLBACK_METADATA_FOLDER, TEMP_FOLDER]);
    const modded = targets.modded;
    const plugins = targets.plugins;
    const patchers = targets.patchers;
    const modInfo = targets.modInfo;
    const pendingDirs: {
        abs: string;
        rel: string;
    }[] = [{ abs: extractRoot, rel: '' }];
    const files: {
        abs: string;
        rel: string;
    }[] = [];
    while (pendingDirs.length > 0) {
        const { abs, rel } = pendingDirs.pop()!;
        let entries: fs.Dirent[];
        try {
            entries = fs.readdirSync(abs, { withFileTypes: true });
        }
        catch {
            continue;
        }
        for (const e of entries) {
            const childAbs = path.join(abs, e.name);
            const childRel = rel ? `${rel}/${e.name}` : e.name;
            if (e.isDirectory()) {
                if (skipNames.has(e.name))
                    continue;
                if (isTemplateFolderName(e.name))
                    continue;
                if (isValidGuidFolderName(e.name)) {
                    if (!modded.some((m) => m.destRel === e.name)) {
                        modded.push({ src: childAbs, destRel: e.name });
                    }
                    continue;
                }
                pendingDirs.push({ abs: childAbs, rel: childRel });
            }
            else {
                files.push({ abs: childAbs, rel: childRel });
            }
        }
    }
    const dllStems = new Set<string>();
    for (const f of files) {
        if (f.rel.toLowerCase().endsWith(PLUGIN_EXT))
            dllStems.add(stemOf(f.rel).toLowerCase());
    }
    for (const f of files) {
        const lower = f.rel.toLowerCase();
        if (!lower.endsWith(PLUGIN_EXT))
            continue;
        if (f.rel.split('/').some((seg) => seg.toLowerCase() === 'patchers')) {
            patchers.push({ src: f.abs, destRel: path.basename(f.rel) });
        }
        else {
            plugins.push({ src: f.abs, destRel: f.rel, extras: [] });
        }
    }
    for (const f of files) {
        const lower = f.rel.toLowerCase();
        const stem = stemOf(f.rel).toLowerCase();
        if (lower.endsWith('.pdb') || lower.endsWith('.xml')) {
            const owner = findPluginEntry(plugins, f.rel, stem);
            owner?.extras.push(f.abs);
            continue;
        }
        if (lower.endsWith('.json')) {
            let stem = lower.slice(0, -'.json'.length);
            const isDeps = stem.endsWith('.deps');
            if (isDeps)
                stem = stem.slice(0, -'.deps'.length);
            if (!dllStems.has(stem))
                continue;
            if (isDeps) {
                const owner = findPluginEntry(plugins, f.rel, stem);
                owner?.extras.push(f.abs);
            }
            else {
                modInfo.push({ src: f.abs, destRel: path.basename(f.rel) });
            }
        }
    }
}
export function collectTargets(input: string): InstallTargets {
    let extractRoot = input;
    const modWrap = path.join(extractRoot, 'Mod');
    if (path.resolve(modWrap) !== path.resolve(extractRoot) &&
        (fs.existsSync(path.join(modWrap, 'BepInEx')) || fs.existsSync(path.join(modWrap, 'BALDI_Data')))) {
        extractRoot = modWrap;
    }
    for (let depth = 0; depth < 5; depth++) {
        const isWrapper = (n: string): boolean => n !== GMP_METADATA_FOLDER && n !== GMP_FALLBACK_METADATA_FOLDER && n !== TEMP_FOLDER;
        const dirs = fs
            .readdirSync(extractRoot, { withFileTypes: true })
            .filter((e) => e.isDirectory() && isWrapper(e.name))
            .map((e) => e.name);
        if (dirs.length !== 1)
            break;
        const child = path.join(extractRoot, dirs[0]);
        const marked = [...pluginsRootCandidates(child), ...moddedRootCandidates(child)].some((p) => fs.existsSync(p));
        if (!marked)
            break;
        extractRoot = child;
    }
    const modded: InstallTargets['modded'] = [];
    const plugins: InstallTargets['plugins'] = [];
    const patchers: InstallTargets['patchers'] = [];
    const modInfo: InstallTargets['modInfo'] = [];
    const moddedRoots = moddedRootCandidates(extractRoot).filter((p) => fs.existsSync(p));
    for (const mr of moddedRoots) {
        for (const e of fs.readdirSync(mr, { withFileTypes: true })) {
            if (e.isDirectory() && !modded.some((x) => x.destRel === e.name)) {
                modded.push({ src: path.join(mr, e.name), destRel: e.name });
            }
        }
    }
    const pluginRoots = pluginsRootCandidates(extractRoot).filter((p) => fs.existsSync(p));
    for (const pr of pluginRoots)
        walkDlls(pr, '', plugins);
    const patchersRoots = patchersRootCandidates(extractRoot).filter((p) => fs.existsSync(p));
    for (const pr of patchersRoots)
        walkPatchers(pr, '', patchers);
    if (pluginRoots.length > 0 || moddedRoots.length > 0 || patchersRoots.length > 0) {
        return { modded, plugins, patchers, modInfo };
    }
    walkHeuristic(extractRoot, { modded, plugins, patchers, modInfo });
    return { modded, plugins, patchers, modInfo };
}
export function deriveModName(targets: InstallTargets): string {
    if (targets.plugins.length > 0) {
        const p = targets.plugins[0].destRel;
        if (p.includes('/'))
            return p.split('/')[0];
        return p.slice(0, -PLUGIN_EXT.length);
    }
    if (targets.modded.length > 0)
        return targets.modded[0].destRel;
    return 'Unknown Mod';
}
export function buildPlan(extractRoot: string): ModInstallPlanDto {
    const targets = collectTargets(extractRoot);
    const modded = [...new Set(targets.modded.map((x) => x.destRel))];
    const plugins = [...new Set(targets.plugins.map((x) => x.destRel))];
    const needsConfirm = !(targets.modded.length > 0 && targets.plugins.length > 0);
    return {
        modName: deriveModName(targets),
        needsConfirm,
        modded,
        plugins
    };
}

File: RBBPMM.Wpf/RBBPMM.Core/ModArchivePlanner.cs
namespace RBBPMM.Core;

public sealed record InstallTarget(string Src, string DestRel);

public sealed record PluginTarget(string Src, string DestRel, List<string> Extras);

public sealed class InstallTargets
{
    public List<InstallTarget> Modded { get; set; } = [];
    public List<PluginTarget> Plugins { get; set; } = [];
    public List<InstallTarget> Patchers { get; set; } = [];
    public List<InstallTarget> ModInfo { get; set; } = [];
}

public sealed class ModInstallPlan
{
    public required string ModName { get; init; }
    public required bool NeedsConfirm { get; init; }
    public required List<string> Modded { get; init; }
    public required List<string> Plugins { get; init; }
}

public static class ModArchivePlanner
{
    private const string PluginExt = ".dll";

    private static string[] PluginsRootCandidates(string extractRoot) =>
    [
        Path.Combine(extractRoot, Constants.BepInExFolder, Constants.PluginsFolder),
        Path.Combine(extractRoot, Constants.PluginsFolder)
    ];

    private static string[] ModdedRootCandidates(string extractRoot) =>
    [
        Path.Combine(extractRoot, Constants.GameDataFolder, "StreamingAssets", "Modded"),
        Path.Combine(extractRoot, "Modded")
    ];

    private static string[] PatchersRootCandidates(string extractRoot) =>
    [
        Path.Combine(extractRoot, Constants.BepInExFolder, Constants.PatcherFolder),
        Path.Combine(extractRoot, Constants.PatcherFolder)
    ];

    private static PluginTarget PluginTarget(string absDll, string rel)
    {
        var dir = Path.GetDirectoryName(absDll) ?? "";
        var stem = Path.GetFileNameWithoutExtension(absDll);
        var extras = new List<string>();
        foreach (var f in new[] { "xml", "pdb" })
        {
            var p = Path.Combine(dir, $"{stem}.{f}");
            if (File.Exists(p))
                extras.Add(p);
        }
        return new PluginTarget(absDll, rel, extras);
    }

    private static void WalkDlls(string dir, string rel, List<PluginTarget> into)
    {
        foreach (var e in Directory.EnumerateFileSystemEntries(dir))
        {
            var name = Path.GetFileName(e);
            var childRel = rel.Length == 0 ? name : $"{rel}/{name}";
            if (Directory.Exists(e))
            {
                WalkDlls(e, childRel, into);
            }
            else if (name.ToLowerInvariant().EndsWith(PluginExt))
            {
                into.Add(PluginTarget(e, childRel));
            }
        }
    }

    private static void WalkPatchers(string dir, string rel, List<InstallTarget> into)
    {
        foreach (var e in Directory.EnumerateFileSystemEntries(dir))
        {
            var name = Path.GetFileName(e);
            var childRel = rel.Length == 0 ? name : $"{rel}/{name}";
            if (Directory.Exists(e))
            {
                WalkPatchers(e, childRel, into);
            }
            else if (name.ToLowerInvariant().EndsWith(PluginExt))
            {
                into.Add(new InstallTarget(e, childRel));
            }
        }
    }

    public static bool IsValidGuidFolderName(string input)
    {
        if (string.IsNullOrEmpty(input))
            return false;
        if (!System.Text.RegularExpressions.Regex.IsMatch(input, @"^[a-z0-9.]+$"))
            return false;
        var segs = input.Split('.');
        return segs.Length is >= 2 and <= 5;
    }

    public static bool IsTemplateFolderName(string input)
    {
        var n = input.ToLowerInvariant();
        return n.Contains("template") || n.Contains("example");
    }

    private static string RelDirOf(string rel)
    {
        var i = rel.LastIndexOf('/');
        return i == -1 ? "" : rel[..i];
    }

    private static string StemOf(string rel)
    {
        var baseName = Path.GetFileName(rel);
        var dot = baseName.LastIndexOf('.');
        return dot == -1 ? baseName : baseName[..dot];
    }

    private static PluginTarget? FindPluginEntry(List<PluginTarget> plugins, string rel, string stem)
    {
        var dir = RelDirOf(rel);
        return plugins.Find(p => RelDirOf(p.DestRel) == dir &&
                                 StemOf(p.DestRel).ToLowerInvariant() == stem)
               ?? plugins.Find(p => StemOf(p.DestRel).ToLowerInvariant() == stem);
    }

    private static void WalkHeuristic(string extractRoot, InstallTargets targets)
    {
        var skipNames = new HashSet<string>
        {
            Constants.GmpMetadataFolder, Constants.GmpFallbackMetadataFolder, Constants.TempFolder
        };
        var pendingDirs = new Stack<(string Abs, string Rel)>();
        pendingDirs.Push((extractRoot, ""));
        var files = new List<(string Abs, string Rel)>();
        while (pendingDirs.Count > 0)
        {
            var (abs, rel) = pendingDirs.Pop();
            IEnumerable<string> entries;
            try
            {
                entries = Directory.EnumerateFileSystemEntries(abs);
            }
            catch (Exception)
            {
                continue;
            }
            foreach (var e in entries)
            {
                var name = Path.GetFileName(e);
                var childRel = rel.Length == 0 ? name : $"{rel}/{name}";
                if (Directory.Exists(e))
                {
                    if (skipNames.Contains(name))
                        continue;
                    if (IsTemplateFolderName(name))
                        continue;
                    if (IsValidGuidFolderName(name))
                    {
                        if (!targets.Modded.Any(m => m.DestRel == name))
                            targets.Modded.Add(new InstallTarget(e, name));
                        continue;
                    }
                    pendingDirs.Push((e, childRel));
                }
                else
                {
                    files.Add((e, childRel));
                }
            }
        }

        var dllStems = new HashSet<string>(StringComparer.OrdinalIgnoreCase);
        foreach (var f in files)
        {
            if (f.Rel.ToLowerInvariant().EndsWith(PluginExt))
                dllStems.Add(StemOf(f.Rel).ToLowerInvariant());
        }

        foreach (var f in files)
        {
            if (!f.Rel.ToLowerInvariant().EndsWith(PluginExt))
                continue;
            if (f.Rel.Split('/').Any(seg => seg.ToLowerInvariant() == "patchers"))
            {
                targets.Patchers.Add(new InstallTarget(f.Abs, Path.GetFileName(f.Rel)));
            }
            else
            {
                targets.Plugins.Add(new PluginTarget(f.Abs, f.Rel, []));
            }
        }

        foreach (var f in files)
        {
            var lower = f.Rel.ToLowerInvariant();
            var stem = StemOf(f.Rel).ToLowerInvariant();
            if (lower.EndsWith(".pdb") || lower.EndsWith(".xml"))
            {
                var owner = FindPluginEntry(targets.Plugins, f.Rel, stem);
                owner?.Extras.Add(f.Abs);
                continue;
            }
            if (lower.EndsWith(".json"))
            {
                var jsonStem = lower[..^".json".Length];
                var isDeps = jsonStem.EndsWith(".deps");
                if (isDeps)
                    jsonStem = jsonStem[..^".deps".Length];
                if (!dllStems.Contains(jsonStem))
                    continue;
                if (isDeps)
                {
                    var owner = FindPluginEntry(targets.Plugins, f.Rel, jsonStem);
                    owner?.Extras.Add(f.Abs);
                }
                else
                {
                    targets.ModInfo.Add(new InstallTarget(f.Abs, Path.GetFileName(f.Rel)));
                }
            }
        }
    }

    public static InstallTargets CollectTargets(string input)
    {
        var extractRoot = input ?? "";
        var modWrap = Path.Combine(extractRoot, "Mod");
        if (Path.GetFullPath(modWrap) != Path.GetFullPath(extractRoot) &&
            (Directory.Exists(Path.Combine(modWrap, Constants.BepInExFolder)) ||
             Directory.Exists(Path.Combine(modWrap, Constants.GameDataFolder))))
        {
            extractRoot = modWrap;
        }

        for (var depth = 0; depth < 5; depth++)
        {
            bool IsWrapper(string n) =>
                n != Constants.GmpMetadataFolder && n != Constants.GmpFallbackMetadataFolder && n != Constants.TempFolder;
            var dirs = Directory.Exists(extractRoot)
                ? Directory.EnumerateFileSystemEntries(extractRoot)
                    .Where(e => Directory.Exists(e) && IsWrapper(Path.GetFileName(e)))
                    .Select(e => Path.GetFileName(e)!)
                    .ToArray()
                : [];
            if (dirs.Length != 1)
                break;
            var child = Path.Combine(extractRoot, dirs[0]);
            var marked = PluginsRootCandidates(child).Concat(ModdedRootCandidates(child)).Any(Directory.Exists);
            if (!marked)
                break;
            extractRoot = child;
        }

        var modded = new List<InstallTarget>();
        var plugins = new List<PluginTarget>();
        var patchers = new List<InstallTarget>();
        var modInfo = new List<InstallTarget>();

        var moddedRoots = ModdedRootCandidates(extractRoot).Where(Directory.Exists).ToArray();
        foreach (var mr in moddedRoots)
        {
            foreach (var e in Directory.EnumerateFileSystemEntries(mr))
            {
                if (Directory.Exists(e))
                {
                    var name = Path.GetFileName(e);
                    if (!modded.Any(x => x.DestRel == name))
                        modded.Add(new InstallTarget(e, name));
                }
            }
        }

        var pluginRoots = PluginsRootCandidates(extractRoot).Where(Directory.Exists).ToArray();
        foreach (var pr in pluginRoots)
            WalkDlls(pr, "", plugins);

        var patchersRoots = PatchersRootCandidates(extractRoot).Where(Directory.Exists).ToArray();
        foreach (var pr in patchersRoots)
            WalkPatchers(pr, "", patchers);

        if (pluginRoots.Length > 0 || moddedRoots.Length > 0 || patchersRoots.Length > 0)
            return new InstallTargets { Modded = modded, Plugins = plugins, Patchers = patchers, ModInfo = modInfo };

        var targets = new InstallTargets { Modded = modded, Plugins = plugins, Patchers = patchers, ModInfo = modInfo };
        WalkHeuristic(extractRoot, targets);
        return targets;
    }

    public static string DeriveModName(InstallTargets targets)
    {
        if (targets.Plugins.Count > 0)
        {
            var p = targets.Plugins[0].DestRel;
            if (p.Contains('/'))
                return p.Split('/')[0];
            return p[..^PluginExt.Length];
        }
        if (targets.Modded.Count > 0)
            return targets.Modded[0].DestRel;
        return "Unknown Mod";
    }

    public static ModInstallPlan BuildPlan(string extractRoot)
    {
        var targets = CollectTargets(extractRoot);
        var modded = targets.Modded.Select(x => x.DestRel).Distinct().ToList();
        var plugins = targets.Plugins.Select(x => x.DestRel).Distinct().ToList();
        var needsConfirm = !(targets.Modded.Count > 0 && targets.Plugins.Count > 0);
        return new ModInstallPlan
        {
            ModName = DeriveModName(targets),
            NeedsConfirm = needsConfirm,
            Modded = modded,
            Plugins = plugins
        };
    }
}
